pkg/scheduler/framework/plugins/podtopologyspread/plugin.go

```diff
@@ -141,9 +144,122 @@ func (pl *PodTopologySpread) EventsToRegister() []framework.ClusterEventWithHint
 		// an unschedulable Pod schedulable.
 		// - Delete. An unschedulable Pod may fail due to violating an existing Pod's topology spread constraints,
 		// deleting an existing Pod may make it schedulable.
-		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.All}},
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.All}, QueueingHintFn: pl.isSchedulableAfterPodChange},
 		// Node add|delete|updateLabel maybe lead an topology key changed,
 		// and make these pod in scheduling schedulable or unschedulable.
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Delete | framework.UpdateNodeLabel}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Delete | framework.UpdateNodeLabel}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
 	}
 }
+
+func (pl *PodTopologySpread) isSchedulableAfterPodChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	originalPod, modifiedPod, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterPodChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.Queue, err
+	}
+
+	constraints, err := pl.getConstraints(pod)
+	if err != nil {
+		logger.Error(err, "get constraints from pod failed", "pod", klog.KObj(pod))
+		return framework.Queue, err
+	}
+
+	// Pod is modified. Return Queue when the label(s) matching topologySpread's selector is added, changed, or deleted.
+	if modifiedPod != nil && originalPod != nil && !reflect.DeepEqual(modifiedPod.Labels, originalPod.Labels) {
+		logger.V(5).Info("pod was updated", "pod", klog.KObj(pod), "modifiedPod", klog.KObj(modifiedPod))
+		// return QueueSkip when the non-related label(s) added/changed/deleted
+		if !(podLabelsMatchSpreadConstraints(constraints, originalPod.Labels) || podLabelsMatchSpreadConstraints(constraints, modifiedPod.Labels)) {
+			return framework.QueueSkip, nil
+		}
+		// return QueueSkip when the pod labels still match constraints after pod updated.
+		if podLabelsMatchSpreadConstraints(constraints, originalPod.Labels) && podLabelsMatchSpreadConstraints(constraints, modifiedPod.Labels) {
```

Comment:
It could be problematic if the Pod has more than one constraint, for example this scenario:
1. Pod-a has two topologyspreadconstraints (constraint-a and constraint-b) and they have a different label selector.
2. The scheduler denied this Pod-a because constraint-a wasn't met. But constraint-b was met actually.
3. There's another Pod named Pod-b, which is already scheduled. It matches with the selector of constraint-b, but doesn't match with the one of constraint-a.
4. Pod-b's label gets updated and now Pod-b matches with both the selector of constraint-a and -b. This change could make Pod-a schedulable because the calculation result of constraint-a could be changed.
5. But this L175 returns QueueSkip and the scheduler doesn't retry Pod-a's scheduling.

Reply:
So, basically we have to change the logic here to check all constraints in both `originalPod` and `modifiedPod` separately.

Reply:
`podLabelsMatchSpreadConstraints` will return true if pod labels match one of constraints selector. Maybe it's wrong.  

> It could be problematic if the Pod has more than one constraint, for example this scenario

This problem could be fixed If `podLabelsMatchSpreadConstraints` return true only if pod labels match all constraints selector. WDYT?

Reply:
> This problem could be fixed If podLabelsMatchSpreadConstraints return true only if pod labels match all constraints selector. WDYT?

I don't think so. When a Pod has two constraints like the above my example, Pod-b actually doesn't need to match with both constraint-a and -b. 

The problematic example:
1. Pod-a has two topologyspreadconstraints (constraint-a and constraint-b) and they have a different label selector.
2. The scheduler denied this Pod-a because constraint-a wasn't met. But constraint-b was met actually.
3. There's another Pod named Pod-b, which is already scheduled. It doesn't match with any selector in Pod-a.
4. Pod-b's label gets updated and now Pod-b matches with the selector of constraint-a. (But, still not match with the selector of constraint-b) This change could make Pod-a schedulable because the calculation result of constraint-a could be changed.
5. But, your new `podLabelsMatchSpreadConstraints` returns false for both originalPod and modifiedPod.

I think the implementation should be something like:

```go
	// Pod is modified. Return Queue when the label(s) matching topologySpread's selector is added, changed, or deleted.
	if modifiedPod != nil && originalPod != nil && !reflect.DeepEqual(modifiedPod.Labels, originalPod.Labels) {
		for _, c := range constraints {
			if c.Selector.Matches(originalPod.Labels) != c.Selector.Matches(modifiedPod.Labels) {
				// This modification makes this Pod match(or not match) with this constraint.
				// Maybe now the scheduling result of topology spread gets changed by this change.
				return framework.Queue, nil
			}
		}
		// This modification of labels doesn't change whether this Pod would match selector or not in any constraints.
		return framework.QueueSkip, nil
	}
```

Reply:
> ```go
> if c.Selector.Matches(originalPod.Labels) != c.Selector.Matches(modifiedPod.Labels) 
> ```

Awesome!